### DISCUSSION TITLE
How can I use daos to filter a record by its relations?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1832] [Q&A]
I have a collection `sites` and it has single relation to `location`.  I want to be able to use daos to run a single query that returns a site if it has a location `name` via a 1-only relation.  Unfortunately I can't figure out how to do such a basic thing.

**Comment by ganigeorgiev** [comment]
You can perform a `JOIN` with the `locations` collection, something like:

```go
func findSitesByLocationName(dao *daos.Dao, name string) ([]*models.Record, error) {
	sitesCollection, err := dao.FindCollectionByNameOrId("sites")
	if err != nil {
		return nil, err
	}

	rows := []dbx.NullStringMap{}

	err = dao.RecordQuery(sitesCollection).
		InnerJoin("locations", dbx.NewExp("locations.id=sites.location")).
		AndWhere(dbx.HashExp{
			"locations.name": name,
		}).
		All(rows)

	if err != nil {
		return nil, err
	}

	return models.NewRecordsFromNullStringMaps(sitesCollection, rows), nil
}
```

**Reply by dschissler** [reply]
This is an easy pattern once seen.  The docs could definitely use an example like this.

You're actually so fast to post that you responded while I was writing about this `RecordQuery` and `NewRecordsFromNullStringMaps` method.

Hey though, you took 17 to 17.5 minutes to respond.  Let's try to keep it under 15.

**Reply by ganigeorgiev** [reply]
The above eventually will be simplified in the future. I'm planning to add auto fail/retry handling for the `SQLITE_BUSY` **read** errors and will create a `RecordsQueryBuilder()` (_or will replace the existing `RecordQuery`, I haven't decided yet_) that will make the above slightly less verbose:

```go
func findSitesByLocationName(dao *daos.Dao, name string) ([]*models.Record, error) {
	var records []*models.Record

	err := dao.RecordQueryBuilder("sites").
		InnerJoin("locations", dbx.NewExp("locations.id=sites.location")).
		AndWhere(dbx.HashExp{
			"locations.name": name,
		}).
		All(records)

	return records, err
}
```

**Reply by dschissler** [reply]
The problem is that if you do `err = dao.RecordQueryBuilder("sites").` then if the collection doesn't exist then you'll have to continue pretending that it does until the end.  I guess that you'd have to wrap dbx methods somehow.

**Reply by satanworker** [reply]
Sorry for basic question, but how to do the same if I have 1 to many relationships?

**Reply by ganigeorgiev** [reply]
@satanworker This will be documented better in the near future, but if a `relation` field has "Max select != 1" then it is stored as json array in the database (eg. `["id1", "id2", "id3"]`). So in this case you can use [`json_each`](https://www.sqlite.org/json1.html#jeach) to perform a join with each individual id, something like:

```go
func findSitesByLocationName(dao *daos.Dao, name string) ([]*models.Record, error) {
	sitesCollection, err := dao.FindCollectionByNameOrId("sites")
	if err != nil {
		return nil, err
	}

	rows := []dbx.NullStringMap{}

	err = dao.RecordQuery(sitesCollection).
		Distinct(true).
		InnerJoin("json_each(sites.locations) je", nil).
		InnerJoin("locations", dbx.NewExp("locations.id=je.value")).
		AndWhere(dbx.HashExp{
			"locations.name": name,
		}).
		All(rows)

	if err != nil {
		return nil, err
	}

	return models.NewRecordsFromNullStringMaps(sitesCollection, rows), nil
}
```

**Reply by dschissler** [reply]
@ganigeorgiev How can I link a "Max select != 1" from collection Site to Point using Dao? I'm simply unfamiliar with sqlite json functions and I'd like to avoid pitfalls and doing it not the Pocketbase way.

**Reply by ganigeorgiev** [reply]
@dschissler I'm not sure that I understand what do you mean.

The previous example with the `json_each` (https://github.com/pocketbase/pocketbase/discussions/1832#discussioncomment-4949125) illustrates how to perform join and look up in a multi-relation field.

If you are asking for how to save a multi-`relation` field value (aka. link relations), you can use the `models.Record.Set("someMultiRelField", []string{"id1", "id2", "id3"})` OR for a custom model, you can use the `types.JsonArray` as struct field (eg. `types.JsonArray{"id1", "id2", "id3"}`).

**Reply by dschissler** [reply]
@ganigeorgiev Thanks I was looking for `model.Record.Set("someMultiRelField", []string{"id1", "id2", "id3"})`.  This is simple.

Do you think that Dao will ever be smart enough to know how to add and delete individual elements / arrays like how the JS SDK can do?

**Reply by ganigeorgiev** [reply]
> Do you think that Dao will ever be smart enough to know how to add and delete individual elements like how the JS SDK can do?

Hm, I don't think this should be part of the `Dao`. Eventually we can add some helper generic slice methods but I think the current behavior is simple and easy enough:

```go
relIds := record.GetStringSlice("someMultiRelField")

// do something with the slice...
// 
// you can also check the go experimental slice utils https://pkg.go.dev/golang.org/x/exp@v0.0.0-20230213192124-5e25df0256eb/slices
// 
// or use some of the PocketBase list helpers - https://pkg.go.dev/github.com/pocketbase/pocketbase@v0.12.3/tools/list
relIds = append(relIds, "newId")

record.Set("someMultiRelField", relIds)

app.Dao().SaveRecord(record)
```

**Reply by dschissler** [reply]
I mean that the Record struct could have that built in to merge or subtract array elements without duplication.
